xc/device.ml

```diff
@@ -1430,6 +1430,107 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
 
 module Dm = struct
 
+  module QMP_Event = struct
+    open Qmp
+    let (pipe_r, pipe_w) = Unix.pipe ()
+
+    let (>>=) m f = match m with | Some x -> f x | None -> ()
+    let (>>|) m f = match m with | Some _ -> () | None -> f ()
+
+    module Lookup = struct
+      let ftod, dtoc = Hashtbl.create 16, Hashtbl.create 16
+      let add c domid =
+        Hashtbl.replace ftod (Qmp_protocol.to_fd c) domid;
+        Hashtbl.replace dtoc domid c
+      let remove c domid =
+        Hashtbl.remove ftod (Qmp_protocol.to_fd c);
+        Hashtbl.remove dtoc domid
+      let domid_of fd = try Some (Hashtbl.find ftod fd) with Not_found -> None
+      let channel_of domid = try Some (Hashtbl.find dtoc domid) with Not_found -> None
+    end
+
+    module Monitor = struct
+      module Epoll = Core.Linux_ext.Epoll
+      module Flags = Core.Linux_ext.Epoll.Flags
+      let create () = (Core.Std.Or_error.ok_exn Epoll.create) ~num_file_descrs:1001 ~max_ready_events:1
+      let wakeup () =  (* write single byte to wake up Monitor.wait *)
+        if Unix.write pipe_w " " 0 1 <> 1 then
+          debug "Pipe write error, failed to wake up qmp event thread"
+      let add m fd = Epoll.set m fd Flags.in_; wakeup ()
+      let remove m fd = Epoll.remove m fd; wakeup ()
+      let wait m = Epoll.wait m ~timeout:`Never
+      let with_event m fn = function
+        | `Ok -> Epoll.iter_ready m ~f:(fun fd flags -> fn fd (flags = Flags.in_))
+        | `Timeout -> debug "Shouldn't receive epoll timeout event in qmp_event_thread"
+    end
+    let m = Monitor.create ()
+
+    let monitor_path domid = Printf.sprintf "/var/run/xen/qmp-event-%d" domid
+    let debug_exn msg e = debug "%s: %s" msg (Printexc.to_string e)
+
+    let remove domid =
+      Lookup.channel_of domid >>= fun c ->
+      finally
```

Comment:
Sorry, I just realised that this can be improved further:

```ocaml
let remove domid =
      Lookup.channel_of domid >>= fun c ->
      try
        finally
          (fun () ->
            Lookup.remove c domid;
            Monitor.remove m (Qmp_protocol.to_fd c);
            debug "Removed QMP Event fd for domain %d" domid)
          (fun () -> Qmp_protocol.close c)
        with e -> debug_exn (Printf.sprintf "Got exception trying to remove qmp on domain-%d" domid) e
```

Reply:
Thanks, updated.